lightning/src/ln/onion_utils.rs

```diff
@@ -1716,8 +1744,15 @@ impl_writeable_tlv_based_enum!(HTLCFailReasonRepr,
 		(_unused, err, (static_value, msgs::OnionErrorPacket { data: data.ok_or(DecodeError::InvalidValue)?, attribution_data })),
 	},
 	(1, Reason) => {
-		(0, failure_code, required),
+		(0, _failure_code, (legacy, u16,
+			|r: &HTLCFailReasonRepr| match r {
+				HTLCFailReasonRepr::LightningError{ .. } => None,
+				HTLCFailReasonRepr::Reason{ failure_reason, .. } => Some(failure_reason.failure_code())
+			})),
 		(2, data, required_vec),
+		// failure_code was required, and is replaced by reason in 0.2 so any time we do not have a
+		// reason available failure_code will be Some and can be expressed as a reason.
+		(4, failure_reason, (default_value, LocalHTLCFailureReason::from(_failure_code.ok_or(DecodeError::InvalidValue)?))),
```

Comment:
Does this work on downgrades with the even type?

Reply:
Right, good catch! Even though it's required now, should be an odd tlv to allow downgrade 👍